coderd/unhanger/detector.go

```diff
@@ -0,0 +1,332 @@
+}
+
+// Error implements error.
+func (e *acquireLockError) Error() string {
+	return "acquire lock: " + e.err.Error()
+}
+
+// Unwrap implements xerrors.Wrapper.
+func (e *acquireLockError) Unwrap() error {
+	return e.err
+}
+
+// Detector automatically detects hung provisioner jobs, sends messages into the
+// build log and terminates them as failed.
+type Detector struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	done   chan struct{}
+
+	db     database.Store
+	pubsub pubsub.Pubsub
+	log    slog.Logger
+	tick   <-chan time.Time
+	stats  chan<- Stats
+}
+
+// Stats contains statistics about the last run of the detector.
+type Stats struct {
+	// HungJobIDs contains the IDs of all jobs that were detected as hung and
+	// terminated.
+	HungJobIDs []uuid.UUID
+	// Error is the fatal error that occurred during the last run of the
+	// detector, if any. Error may be set to AcquireLockError if the detector
+	// failed to acquire a lock.
+	Error error
+}
+
+// New returns a new hang detector.
+func New(ctx context.Context, db database.Store, pub pubsub.Pubsub, log slog.Logger, tick <-chan time.Time) *Detector {
+	//nolint:gocritic // Hang detector has a limited set of permissions.
+	ctx, cancel := context.WithCancel(dbauthz.AsHangDetector(ctx))
+	le := &Detector{
+		ctx:    ctx,
+		cancel: cancel,
+		done:   make(chan struct{}),
+		db:     db,
+		pubsub: pub,
+		log:    log,
+		tick:   tick,
+		stats:  nil,
+	}
+	return le
+}
+
+// WithStatsChannel will cause Executor to push a RunStats to ch after
+// every tick. This push is blocking, so if ch is not read, the detector will
+// hang. This should only be used in tests.
+func (d *Detector) WithStatsChannel(ch chan<- Stats) *Detector {
+	d.stats = ch
+	return d
+}
+
+// Start will cause the detector to detect and unhang provisioner jobs on every
+// tick from its channel. It will stop when its context is Done, or when its
+// channel is closed.
+//
+// Start should only be called once.
+func (d *Detector) Start() {
+	go func() {
+		defer close(d.done)
+		defer d.cancel()
+
+		for {
+			select {
+			case <-d.ctx.Done():
+				return
+			case t, ok := <-d.tick:
+				if !ok {
+					return
+				}
+				stats := d.run(t)
+				if stats.Error != nil && !xerrors.As(stats.Error, &acquireLockError{}) {
+					d.log.Warn(d.ctx, "error running workspace build hang detector once", slog.Error(stats.Error))
+				}
+				if len(stats.HungJobIDs) != 0 {
+					d.log.Warn(d.ctx, "detected (and terminated) hung provisioner jobs", slog.F("job_ids", stats.HungJobIDs))
+				}
+				if d.stats != nil {
+					select {
+					case <-d.ctx.Done():
+						return
+					case d.stats <- stats:
+					}
+				}
+			}
+		}
+	}()
+}
+
+// Wait will block until the detector is stopped.
+func (d *Detector) Wait() {
+	<-d.done
+}
+
+// Close will stop the detector.
+func (d *Detector) Close() {
+	d.cancel()
+	<-d.done
+}
+
+func (d *Detector) run(t time.Time) Stats {
+	ctx, cancel := context.WithTimeout(d.ctx, 5*time.Minute)
```

Comment:
In the event that we would have 100 hung jobs, a per-job transaction (see other comment) would also help ensure that this context is never too short to at least do _some_ work (and eventually mark everything).

Reply:
That would be nice, but having multiple transactions while holding a lock is not possible with the in-memory DB. I tried doing multiple transactions at first but hit this issue. Adding multiple transaction support to in-memory DB seems like a bad idea because we don't have table/row locks like PSQL.

Reply:
I might just add a limit of 10 jobs per run so we don't timeout and risk rolling back.